src/core/errorHandler.js

```diff
@@ -0,0 +1,79 @@
+import React, { PropTypes } from 'react';
+import { compose } from 'redux';
+import { connect } from 'react-redux';
+
+import { clearError, setError } from 'core/actions/errors';
+import log from 'core/logger';
+
+import 'core/css/ErrorHandler.scss';
+
+function generateHandlerId({ name = '' } = {}) {
+  return `${name}-${Math.random().toString(36).substr(2, 9)}`;
+}
+
+export class ErrorHandler {
+  constructor({ id, dispatch }) {
+    this.id = id;
+    this.dispatch = dispatch;
+  }
+
+  clear() {
+    log.info('Clearing last error for ', this.id);
+    this.dispatch(clearError(this.id));
+  }
+
+  handle(error) {
+    const info = { error, id: this.id };
+    log.info('Dispatching error action', info);
+    this.dispatch(setError(info));
+  }
+}
+
+export function withErrorHandling({
+  name, id = generateHandlerId({ name }),
+} = {}) {
+  return (WrappedComponent) => {
+    class ErrorHandlerComponent extends React.Component {
+      static propTypes = {
+        error: PropTypes.object,
+        dispatch: PropTypes.func,
+      }
+
+      render() {
+        const { error, dispatch, ...otherProps } = this.props;
+
+        const errorHandler = new ErrorHandler({ id, dispatch });
+        const props = { ...otherProps, errorHandler };
+
+        let errorInfo;
+        if (error) {
+          let errorMessages = [];
+          errorMessages = error.messages.map((msg) => <li>{msg}</li>);
+          errorInfo = (
+            <ul className="ErrorHandler-list">
+              {errorMessages}
+            </ul>
+          );
+        }
+
+        return (
+          <div>
+            {errorInfo}
+            <WrappedComponent {...props} />
```

Comment:
This looks like a more flexible approach but it would also add more boilterplate. I think the behavior of displaying an error at the top would be adequate for most components so what I'd like to do is proceed with the current approach and expand it to your more flexible approach later as an option in a future patch. It could be done maybe more like `withErrorHandling({ manageErrorView: false, ... })` for components that want to manage the view themselves.